packages/stan/src/derived.js

```diff
@@ -0,0 +1,199 @@
+/**
+ * External dependencies
+ */
+import { noop } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { createQueue } from '@wordpress/priority-queue';
+
+const resolveQueue = createQueue();
+
+/** @typedef {import('./types').WPCommonAtomConfig} WPCommonAtomConfig */
+/**
+ * @template T
+ * @typedef {import("./types").WPDerivedAtomResolver<T>} WPDerivedAtomResolver
+ */
+/**
+ * @template T
+ * @typedef {import("./types").WPDerivedAtomUpdater<T>} WPDerivedAtomUpdater
+ */
+/**
+ * @template T
+ * @typedef {import("./types").WPAtom<T>} WPAtom
+ */
+/**
+ * @template T
+ * @typedef {import("./types").WPAtomState<T>} WPAtomState
+ */
+
+/**
+ * Creates a derived atom.
+ *
+ * @template T
+ * @param {WPDerivedAtomResolver<T>} resolver Atom resolver.
+ * @param {WPDerivedAtomUpdater<T>}  updater  Atom updater.
+ * @param {WPCommonAtomConfig=}      config   Common Atom config.
+ * @return {WPAtom<T>} Created atom.
+ */
+export const createDerivedAtom = ( resolver, updater = noop, config = {} ) => (
+	registry
+) => {
+	/**
+	 * @type {any}
+	 */
+	let value = null;
+
+	/**
+	 * @type {Set<() => void>}
+	 */
+	const listeners = new Set();
+
+	/**
+	 * @type {(WPAtomState<any>)[]}
+	 */
+	let dependencies = [];
+	let isListening = false;
+	let isResolved = false;
+	const context = {};
+
+	const dependenciesUnsubscribeMap = new WeakMap();
+
+	const notifyListeners = () => {
+		listeners.forEach( ( l ) => l() );
+	};
+
+	const refresh = () => {
+		if ( listeners.size > 0 ) {
+			if ( config.isAsync ) {
+				resolveQueue.add( context, resolve );
+			} else {
+				resolve();
+			}
+		} else {
+			isListening = false;
+		}
+	};
+
+	/**
+	 * @param {WPAtomState<any>} atomState
+	 */
+	const addDependency = ( atomState ) => {
+		if ( ! dependenciesUnsubscribeMap.has( atomState ) ) {
+			dependenciesUnsubscribeMap.set(
+				atomState,
+				atomState.subscribe( refresh )
+			);
+		}
+	};
+
+	const resolve = () => {
+		/**
+		 * @type {(WPAtomState<any>)[]}
+		 */
+		const updatedDependencies = [];
+		const updatedDependenciesMap = new WeakMap();
+		let result;
+		const unresolved = [];
+		try {
+			result = resolver( {
+				get: ( atom ) => {
+					const atomState = registry.__unstableGetAtomState( atom );
+					// It is important to add the dependency as soon as it's used
+					// because it's important to retrigger the resolution if the dependency
+					// changes before the resolution finishes.
+					addDependency( atomState );
+					updatedDependenciesMap.set( atomState, true );
+					updatedDependencies.push( atomState );
+					if ( ! atomState.isResolved ) {
+						unresolved.push( atomState );
+					}
+					return atomState.get();
+				},
+			} );
+		} catch ( error ) {
+			if ( unresolved.length === 0 ) {
+				throw error;
```

Comment:
If I read it correctly, it will silence any exceptions during the resolution.

Reply:
I think the entire try/catch could be avoided if `get: ( atom ) => {` was async and defered the responsibility of waiting for dependencies to `resolver` function. I'll propose a change later today.

Reply:
(or there's always [this functional API](https://github.com/WordPress/gutenberg/pull/26866/files#r527058918) idea where get: (atom) would return an observable and reduce this file to ~20 LOC)

Reply:
how so? It throws the error once again?

Reply:
Only if `unresolved.length === 0`, I imagine there could be a case where dependent atom's `resolve()` throws an error in this line: `return atomState.get();` and this error gets supressed since `unresolved.length` is no longer 0. Although the solution could be maybe as simple as moving `.get()` above push.

Reply:
Catching these is intended :). If we have an unresolved value, we might have unexpected exceptions because the  resolver might expect a number but since it's unresolved, it will have `null`, so maybe the atom author will use it in an addition and this will trigger an exception until the value is resolved. That's fine but if there's really an error in the logic, the error will be thrown once all the values are resolved.

Reply:
@youknowriad alright, I was right after all - here's a test case where the error is not thrown:
```js
	it( 'should not suppress resolution errors', async () => {
		const count1 = createAtom( 1 );
		const asyncCount = createDerivedAtom( async ( { get } ) => {
			throw new Error();
		} );
		const sum = createDerivedAtom( async ( { get } ) => {
			return get( count1 ) + get( asyncCount );
		} );
		const registry = createAtomRegistry();
		expect( () => {
			registry.get( sum ); // or registry.subscribe( sum, () => {} );
		} ).toThrow();
	} );
```

I played with adding `flushImmediatesAndTicks` but it doesn't help. Either way, node logs `(node:31681) UnhandledPromiseRejectionWarning: Error` as `resolve()` is not awaited for nor are its resolution errors caught. I'm playing with a small refactor to address that problem.

Reply:
This case is very weird :P While I agree we should have better error handling, I'm not sure we should be throwing errors because in the example above, you have a failing promise and not an exception that is thrown

Reply:
The `sum` resolver with the `+` operation is special because `+` doesn't mind if one argument is `null`:
```js
42 + null === 42;
null + null === 0;
```
The `sum` resolver will never throw. It will return questionable sums instead 🙂 

It's better to test this with a resolver that doesn't tolerate `null` values. Like when the upstream values are arrays.

Reply:
@jsnajdr IMHO that should be two separate tests - sum resolver is also a valid use case.

Reply:
I couldn't come up with anything that wouldn't `async/await` all the things, here's my attempt: https://github.com/WordPress/gutenberg/pull/27155